Topics/Tech_Stacks/Multithreading.md

```diff
@@ -0,0 +1,177 @@
+# Multithreading
+
+## Concepts
+- Motivation for multithreading
+	- Responsiveness achieved by concurrency(multi-tasking between threads)
+	- Performance achieved by parallelism(multiple tasks executing in parallel)
+
+- The thread contains
+	- Stack: Region in memory, where local variables are stored, and passed into functions
+	- Instruction Pointer: Address of the next instruction to execute
+- Each thread is executing a different instruction and a different function at any given moment
+
+- The threads share
+	- Files
+	- Heap
+	- Code
+
+- Context Switch
+	- Each process has multiple threads
+		- Stop thread 1
+		- Schedule thread 1 out
+		- Schedule thread 2 in
+		- Start Thread 2
+	- Each thread consumes resources in the CPU and memory
+	- When we switch to a different thread
+		- Store data for one thread
+		- Restore data for another thread
+
+	1. Too many threads will cause thrashing, spending more time in management than real productive work
+	2. Threads consume less resources than processes
+	3. Context switching between threads from the same process is cheaper than context switch between different processes
+
+- Thread Scheduling
+	- Operating system divides the time into moderately sized pieces called epochs
+	- In each epochs, the operating system allocates a different time slice for each thread
+	- Not all the threads get to run or complete in each epoch
+	- The decision is based on Dynamic Priority = Static Priority + Bonus(can be negative)
+		- Static Priority is set be the developer programmatically
+		- Bonus is adjusted by the operating system in every epoch, for each thread
+	- Using Dynamic Priority, the OS will give preference for interactive threads and threads that did not complete in the last epochs, or did not get enough time to run, preventing Starvation
+
+---
+
+## Thread and Process
+- Multi-Thread
+	- Multithreaded if the tasks share a lot of data
+	- Threads are much faster to create an destroy
+	- Switching between threads of the same process is faster(shorter context switches)
+- Multi-Process
+	- Security and stability are of higher importance
+	- Tasks are unrelated to each other
+
+---
+
+## Tread Class
```

Comment:
I find the concept of the Thread Class a bit unclear. Could you kindly provide some code examples to illustrate how to use it?